Rework rustdoc const type

GuillaumeGomez opened the pull request:
This PR is mostly about two things:
 1. Not storing some information in the `clean::Constant` type
 2. Using `TyCtxt` in the formatting (which we will need in any case as we move forward in any case).

Also: I'm very curious of the perf change in here.

Thanks a lot @danielhenrymantilla for your `Captures` idea! It allowed me to solve the lifetime issue completely. :)

r? @jyn514

GuillaumeGomez commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit b9c09e42b696c2dc58090555fd7e67592ff2e8af with merge b9ff6ae773d5f0ce233ef2dca5abb531e05c6809...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b9c09e42b696c2dc58090555fd7e67592ff2e8af","merge_sha":"b9ff6ae773d5f0ce233ef2dca5abb531e05c6809"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2051856421)
Build commit: b9ff6ae773d5f0ce233ef2dca5abb531e05c6809 (`b9ff6ae773d5f0ce233ef2dca5abb531e05c6809`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2051856421"},"merge_sha":"b9ff6ae773d5f0ce233ef2dca5abb531e05c6809"} -->

rust-timer commented:
Queued b9ff6ae773d5f0ce233ef2dca5abb531e05c6809 with parent 66ec64ccf31883cd2c28d045912a76179c0c6ed2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=66ec64ccf31883cd2c28d045912a76179c0c6ed2&end=b9ff6ae773d5f0ce233ef2dca5abb531e05c6809).

jyn514 commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1745,12 +1743,7 @@ impl<'tcx> Clean<Type> for Ty<'tcx> {
-            value: None,
-            is_literal: false,
-        }
+        Constant::TyConst { type_: self.ty.clean(cx), expr: format!("{}", self) }
```
Rather than stringifying `self`, could you store it in the `Constant` and only stringify on demand? Or does that make the lifetimes painful?

GuillaumeGomez replied:
It's worth giving it a try.

GuillaumeGomez replied:
So, after taking a quick look, I propose this instead: I'll add a FIXME so we did it in another PR because this will otherwise just grow the diff once again, and it's already huge.

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,58 @@ crate struct Static {
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    TyConst { type_: Type, expr: String },
+    Generic { type_: Type, body: BodyId },
+    Inline { type_: Type, did: DefId },
+    Const { type_: Type, did: DefId, body: BodyId },
+}
```
Can you document what these are? I can guess `Inline` is for `const { ... }` blocks, but I can't tell about the rest.

GuillaumeGomez replied:
Arf sorry, I was experimenting on it for so long that I didn't document things...

jyn514 commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1052,11 +1053,10 @@ fn render_assoc_item(
-                .print(cx.cache()),
-            spotlight = spotlight_decl(&d, cx.cache()),
-            where_clause = WhereClause { gens: g, indent, end_newline }.print(cx.cache())
+            generics = g.print(cx.cache(), cx.tcx()),
+            decl = d.full_print(cx.cache(), cx.tcx(), header_len, indent, header.asyncness),
+            spotlight = spotlight_decl(&d, cx.cache(), cx.tcx()),
+            where_clause = print_where_clause(g, cx.cache(), cx.tcx(), indent, end_newline),
```
You use `tcx()` quite a lot - maybe assign it to a temporary?

Note that this will have pretty big conflicts with https://github.com/rust-lang/rust/pull/82855.

GuillaumeGomez replied:
I'll rebase then. ;)

jyn514 reviewed (commented):
I haven't read the whole thing yet. Could you describe more the motivation for this change? You said

> Not storing some information in the clean::Constant type

which makes sense to me - is the idea just to help performance?

rust-timer commented:
Finished benchmarking try commit (b9ff6ae773d5f0ce233ef2dca5abb531e05c6809): [comparison url](https://perf.rust-lang.org/compare.html?start=66ec64ccf31883cd2c28d045912a76179c0c6ed2&end=b9ff6ae773d5f0ce233ef2dca5abb531e05c6809).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

GuillaumeGomez commented:
The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):

![Screenshot from 2021-03-07 22-37-53](https://user-images.githubusercontent.com/3050060/110255842-1ccdbf80-7f96-11eb-8591-8eca0d765d53.png)

GuillaumeGomez commented:
> I haven't read the whole thing yet. Could you describe more the motivation for this change? You said
> 
> > Not storing some information in the clean::Constant type
> 
> which makes sense to me - is the idea just to help performance?

It's definitely targetting performance here. It's also preparing the field in `html/format` by passing the `TyCtxt` type around since we'll very likely make more things into the formatting directly too (and this is where most changes actually come from...).

GuillaumeGomez commented:
So let's see if your idea will improve the perf, even a little. ;)

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 4357e647282efb170f5ada5d43e09ae23eb62b7f with merge d1670c5e02aa01a73205d1a1ff58e5fda1044425...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4357e647282efb170f5ada5d43e09ae23eb62b7f","merge_sha":"d1670c5e02aa01a73205d1a1ff58e5fda1044425"} -->

jyn514 commented:
> It's also preparing the field in html/format by passing the TyCtxt type around since we'll very likely make more things into the formatting directly too (and this is where most changes actually come from...).

Ok, so the TyCtxt isn't actually needed for these changes? Could you split it into a different PR (or just commit) so it's easier to see which changes are necessary?

GuillaumeGomez commented:
Well... Actually it is needed because you need it to print the const. And from that, all the "upper" functions needed it too. You'll note that not all functions in `html/format` take a `TyCtxt`, that's the reason. I "only" changed the ones which needed it because one of the children was using it. Pretty impressive how one change is affecting so many things. If it wasn't needed yet, I wouldn't have been able to add a parameter without using underscore to prevent warnings. ;)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2052459732)
Build commit: d1670c5e02aa01a73205d1a1ff58e5fda1044425 (`d1670c5e02aa01a73205d1a1ff58e5fda1044425`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2052459732"},"merge_sha":"d1670c5e02aa01a73205d1a1ff58e5fda1044425"} -->

rust-timer commented:
Queued d1670c5e02aa01a73205d1a1ff58e5fda1044425 with parent 66ec64ccf31883cd2c28d045912a76179c0c6ed2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=66ec64ccf31883cd2c28d045912a76179c0c6ed2&end=d1670c5e02aa01a73205d1a1ff58e5fda1044425).

rust-timer commented:
Finished benchmarking try commit (d1670c5e02aa01a73205d1a1ff58e5fda1044425): [comparison url](https://perf.rust-lang.org/compare.html?start=66ec64ccf31883cd2c28d045912a76179c0c6ed2&end=d1670c5e02aa01a73205d1a1ff58e5fda1044425).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

GuillaumeGomez commented:
No notable changes since before. Waiting for the review then!

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// Typed constant value.
```
What does this mean? Aren't all values typed?

GuillaumeGomez replied:
I took the documentation from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Const.html

If you have ideas for improvements, it's very welcome. :laughing:

camelid replied:
I \*think\* it might be referring to the fact that [`ConstValue`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/value/enum.ConstValue.html) is basically loosely-structured raw bytes and you need the type to figure out what it actually means.

jyn514 replied:
> I took the documentation from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Const.html

Ah ok, you literally mean `ty::Const`, I thought it had some deeper meaning. And the reason it can't be combined with `Local` is because non-local constants won't have a `BodyId`. Can you add both those things as a comment? And please remove "A typed constant value" - that would be good documentation for `enum Constant`, but it's not helpful for this variant because it's true for *all* the `Constant` variants.

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Generic { type_: Type, body: BodyId },
```
Generic seems like a bad name for this. Maybe "Anonymous" is better?

GuillaumeGomez replied:
The original type is [AnonConst](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.AnonConst.html) so you're right, it'd be much better!

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Generic { type_: Type, body: BodyId },
+    /// Inlined constant (from another crate).
+    Inline { type_: Type, did: DefId },
```
Constants can be inlined without being from another crate. I think `Extern` or `ExternItem` would be a better name.

GuillaumeGomez replied:
Let's go for `Extern` then.

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+    /// Inlined constant (from another crate).
+    Inline { type_: Type, did: DefId },
+    /// const FOO: u32 = ...;
+    Const { type_: Type, did: DefId, body: BodyId },
```
"Const" is not very descriptive. Maybe "Local" or "Item"?

GuillaumeGomez replied:
Let's go for `Local` then!

jyn514 reviewed (commented):
> The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):

Wow, that's amazing! I looked at the self-profile data and it looks like we've almost completely eliminated the time `eval_to_allocation_raw` takes to execute (!!). Do you know how that happened? I don't see it being called in rustdoc directly.

> eval_to_allocation_raw | 0.002 | 0.39% | -99.9% (-1.191) | 151 | -5.0% (-8.000)

I haven't looked at the rest of the changes in detail but this definitely seems worth landing if it has that big an impact :)

GuillaumeGomez commented:
> > The benchmark results seem mostly positive (just wondering what's happening with `ctfe-stress-4-doc`):
> 
> Wow, that's amazing! I looked at the self-profile data and it looks like we've almost completely eliminated the time `eval_to_allocation_raw` takes to execute (!!). Do you know how that happened? I don't see it being called in rustdoc directly.

I assume that it was called indirectly, as most things. But to be honest, I didn't expect this to have such a great impact. :)

> > eval_to_allocation_raw | 0.002 | 0.39% | -99.9% (-1.191) | 151 | -5.0% (-8.000)
> 
> I haven't looked at the rest of the changes in detail but this definitely seems worth landing if it has that big an impact :)

Awesome! Fixing your comments then as quickly as possible then!

bors commented:
:umbrella: The latest upstream changes (presumably #82896) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

GuillaumeGomez commented:
And conflicts fixed!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2059223997) [(plain)](https://github.com/rust-lang/rust/commit/141b1c9de4b7dd83a5a14378d81569ed773a9a2b/checks/2059223997/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustdoc/json/mod.rs at line 108:
                                 .last()
                                 .map(Clone::clone),
                             visibility: types::Visibility::Public,
-                            inner: types::ItemEnum::Trait(
-                                trait_item.clone().into_tcx(self.tcx),
-                            ),
+                            inner: types::ItemEnum::Trait(trait_item.clone().into_tcx(self.tcx)),
                             source: None,
                             docs: Default::default(),
                             links: Default::default(),
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/error.rs" "/checkout/src/librustdoc/json/mod.rs" "/checkout/src/librustdoc/config.rs" "/checkout/src/librustdoc/visit_lib.rs" "/checkout/src/librustdoc/visit_ast.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/lib.rs" "/checkout/src/librustdoc/doctest/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:18
```

</details>

GuillaumeGomez commented:
Fixed format issue.

jyn514 commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1135,7 +1133,7 @@ impl Clean<Item> for ty::AssocItem {
                 let ty = tcx.type_of(self.def_id);
                 let default = if self.defaultness.has_value() {
-                    Some(inline::print_inlined_const(cx, self.def_id))
+                    Some(inline::print_inlined_const(cx.tcx, self.def_id))
```
```suggestion
                    Some(inline::print_inlined_const(tcx, self.def_id))
```

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+        }
+    }
+
+    crate fn type_(&self) -> &Type {
+        match *self {
+            Self::TyConst { ref type_, .. }
+            | Self::Extern { ref type_, .. }
+            | Self::Local { ref type_, .. }
+            | Self::Anonymous { ref type_, .. } => type_,
+        }
+    }
+
+    crate fn to_type(self) -> Type {
+        match self {
+            Self::TyConst { type_, .. }
+            | Self::Extern { type_, .. }
+            | Self::Local { type_, .. }
+            | Self::Anonymous { type_, .. } => type_,
+        }
+    }
```
Have you considered using a struct that holds the type instead? Then you wouldn't need these two functions and you could just access the type directly:
```rust
struct Constant {
  type_: Type,
  kind: ConstantKind,
}

enum ConstantKind { ... }
```

GuillaumeGomez replied:
I did but I wondered for a while if it was worth it or not. I personally think that the current code is fine, but if you prefer otherwise, I can update.

jyn514 replied:
I think I would rather have the explicit field, yeah.

GuillaumeGomez replied:
Then let's go for it!

jyn514 commented on src/librustdoc/html/format.rs:
```diff
@@ -107,31 +107,6 @@ impl Buffer {
     }
 }
 
-/// Wrapper struct for properly emitting a function or method declaration.
-crate struct Function<'a> {
-    /// The declaration to emit.
-    crate decl: &'a clean::FnDecl,
-    /// The length of the function header and name. In other words, the number of characters in the
-    /// function declaration up to but not including the parentheses.
-    ///
-    /// Used to determine line-wrapping.
-    crate header_len: usize,
-    /// The number of spaces to indent each successive line with, if line-wrapping is necessary.
-    crate indent: usize,
-    /// Whether the function is async or not.
-    crate asyncness: hir::IsAsync,
-}
-
-/// Wrapper struct for emitting a where-clause from Generics.
-crate struct WhereClause<'a> {
-    /// The Generics from which to emit a where-clause.
-    crate gens: &'a clean::Generics,
-    /// The number of spaces to indent each line with.
-    crate indent: usize,
-    /// Whether the where-clause needs to add a comma and newline after the last bound.
-    crate end_newline: bool,
-}
```
Where did these go?

GuillaumeGomez replied:
Removed and replaced by functions.

GuillaumeGomez replied:
To be more precise: there was a huge lifetime issue so I fixed it by simply removing the two structs.

jyn514 commented on src/librustdoc/html/format.rs:
```diff
@@ -966,147 +1012,176 @@ impl clean::BareFunctionDecl {
-                if i == 0 {
-                    args.push_str("<br>");
-                }
+    /// * `header_len`: The length of the function header and name. In other words, the number of
+    ///   characters in the function declaration up to but not including the parentheses.
+    ///   <br>Used to determine line-wrapping.
+    /// * `indent`: The number of spaces to indent each successive line with, if line-wrapping is
+    ///   necessary.
+    /// * `asyncness`: Whether the function is async or not.
+    crate fn full_print<'b, 'a: 'b, 'tcx: 'a>(
+        &'a self,
+        cache: &'b Cache,
+        tcx: TyCtxt<'tcx>,
+        header_len: usize,
+        indent: usize,
+        asyncness: hir::IsAsync,
+    ) -> impl fmt::Display + 'b + Captures<'tcx> {
+        display_fn(move |f| self.inner_full_print(cache, tcx, header_len, indent, asyncness, f))
```
Why is this new function needed?

GuillaumeGomez replied:
To be more precise: there was a huge lifetime issue so I fixed it by simply removing the two structs.

jyn514 commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1172,21 +1181,19 @@ fn render_assoc_items(
                 let id = cx.derive_id(small_url_encode(format!(
                     "deref-methods-{:#}",
-                    type_.print(cx.cache())
+                    type_.print(cache, tcx)
```
Are all these changes from `cx.cache()` to `cache` necessary? They're not *bad*, but they make an already giant diff even bigger, it would be nice to put them in a separate commit or something.

jyn514 commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1582,7 +1595,16 @@ fn render_impl(
+                    default.as_ref(),
+                    link.anchor(&id),
+                    "",
+                    cx.cache(),
+                    tcx,
+                );
```
Rather than passing both of these, could you just pass a `cx`?

GuillaumeGomez replied:
Same: I can, but then I'd have to call again both `cx.cache()` and `cx.tcx()`. So as you prefer?

jyn514 replied:
Nah, this is fine.

jyn514 commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -1582,7 +1595,16 @@ fn render_impl(
+                    default.as_ref(),
+                    link.anchor(&id),
+                    "",
+                    cx.cache(),
+                    tcx,
+                );
```
Rather than passing both of these, could you just pass a `cx`?

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+            Self::TyConst { .. } => false,
+            Self::Inline { did, .. } => did
+                .as_local()
+                .map_or(false, |did| is_literal_expr(tcx, tcx.hir().local_def_id_to_hir_id(did))),
```
I took this code from here: https://github.com/rust-lang/rust/blob/4d7a6484852603749f765479ee7288e6ae7186f4/src/librustdoc/clean/inline.rs#L507-L509

I didn't change its behavior.

jyn514 replied:
Hmm, ok - it doesn't need to change here then.

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+    /// used to define explicit discriminant values for enum variants.
+    Generic { type_: Type, body: BodyId },
+    /// Inlined constant (from another crate).
+    Inline { type_: Type, did: DefId },
```
Sure!

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,64 @@ crate struct Static {
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, did: DefId },
```
This one uses `print_inlined_const` and is generated from `clean/inline.rs` whereas `TyConst` simply uses `to_string()` from its expression.

jyn514 replied:
Right, sure, but why do you need it to be separate? Could you store this instead wherever you store a `TyConst`?

GuillaumeGomez replied:
Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. :laughing: 

Do you want me to add a FIXME so we can do it in another PR?

jyn514 replied:
> Because it's a reference and adding a lifetime everywhere seemed like a bit too much for this PR. laughing

Can't you just store the `DefId` instead of storing the whole `ty::Const`?

GuillaumeGomez replied:
The `ty::Const` comes from https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/subst/enum.GenericArgKind.html so I wonder if `tcx.type_of(def_id)` will actually allow me to get it without too much troubles.

The easiest way seem to simply remove the clean impl for `ty::Const` and instead take it by value into the enum. However, it has a lifetime which is what we don't want to do here if we want to avoid a super big diff. I already added a FIXME, so unless there is a way to get the `ty::Const` from the `DefId`, I think it's better to keep it this way here and then move the `ty::Const` directly into the enum in another PR (and with all the lifetime changes it'll bring).

However, if you know a way to get the `ty::Const`, I'd be more than happy to do it here directly. I just couldn't find a way to get it though...

jyn514 replied:
@GuillaumeGomez why did you mark this as resolved? I vaguely remember we talked about this in DMs but I don't remember what we decided - I still feel like TyConst could be combined into `Extern` by only storing the DefId.

jyn514 replied:
I guess my question more broadly is why you need the `ty::Const` for `TyConst` when you don't need it for `Extern`.

GuillaumeGomez replied:
Because `ty::Const` is a generic, not a type and can't be retrived easily just from its `DefId`. Which is why you said on zulip that it was fine to keep as is.

jyn514 replied:
Thanks! That was it :)

jyn514 commented on src/librustdoc/clean/mod.rs:
```diff
@@ -1745,12 +1743,8 @@ impl<'tcx> Clean<Type> for Ty<'tcx> {
-            value: None,
-            is_literal: false,
-        }
+        // FIXME: instead of storing `format!("{}", self)`, store `self` directly instead.
```
```suggestion
        // FIXME: instead of storing the stringified expression, store `self` directly instead.
```

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,67 @@ crate struct Static {
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, def_id: DefId },
```
This is specifically for non-local constants I thought? Or does `clean::inline` discard the BodyId? That doesn't seem right.
```suggestion
    /// A constant from a different crate.
    Extern { type_: Type, def_id: DefId },
```

jyn514 commented on src/librustdoc/clean/types.rs:
```diff
@@ -1966,11 +1967,67 @@ crate struct Static {
+        }
+    }
+
+    crate fn value(&self, tcx: TyCtxt<'_>) -> Option<String> {
```
What does 'value' mean? Why do only `Extern` and `Local` constants have one?

GuillaumeGomez replied:
It's the `= "whatever";` part. I took the "value" name from the field of the original `clean::Constant` struct.

jyn514 replied:
How is it different from 'expr'?

GuillaumeGomez replied:
That's actually a good question now that I think about it. :laughing:

jyn514 replied:
Can you try switching all uses of `value()` to use `expr()` instead and see if it breaks anything? If not, could you remove this function?

GuillaumeGomez replied:
It's worth a try.

GuillaumeGomez replied:
So I think it's worth thinking about it later on because:

```rust
let value = c.value(cx.tcx());
let is_literal = c.is_literal(cx.tcx());
let expr = c.expr(cx.tcx());
if value.is_some() || is_literal {
    write!(w, " = {expr};", expr = Escape(&expr));
} else {
    w.write_str(";");
}

if !is_literal {
    if let Some(value) = &value {
        let value_lowercase = value.to_lowercase();
        let expr_lowercase = expr.to_lowercase();

        if value_lowercase != expr_lowercase
            && value_lowercase.trim_end_matches("i32") != expr_lowercase
        {
            write!(w, " // {value}", value = Escape(value));
        }
    }
}
```

Both `expr` and `value` are used.

jyn514 replied:
That doesn't really answer my question though  - why not just unconditionally print the expression, instead of printing it for some types of constants not others?

GuillaumeGomez replied:
It needs to be investigated. But I don't think it's a good idea to do it here. Should I open an issue about this?

jyn514 replied:
Hmm, ok, we can look into it in a follow-up. If you open an issue that would be great.

GuillaumeGomez commented:
Split the enum into a struct and an enum. :)

GuillaumeGomez commented:
@camelid Updated the documentation. :)

jyn514 commented:
This is waiting on https://github.com/rust-lang/rust/pull/82873#discussion_r591818612, and then I need to read the full diff.

camelid commented on src/librustdoc/clean/types.rs:
```diff
@@ -1968,9 +1969,55 @@ crate struct Static {
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+crate enum ConstantKind {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { expr: String },
+    /// A constant (expression) that's not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { body: BodyId },
+    /// A constant from a different crate.
+    Extern { def_id: DefId },
```
I feel confused about this: `TyConst` says (emphasis mine)

> This is the wrapper around `ty::Const` for a **non-local constant**.

but `Extern` *also* says

> A constant from a **different crate**.

What's the difference between a "non-local constant" and a "constant from a different crate"?

GuillaumeGomez replied:
`TyConst` is about generics (meaning, "non local") whereas `Extern` is for foreign constants items.

camelid replied:
When you say "about generics" do you mean it's about const generics? What's an example of a `TyConst` and of an `Extern`?

GuillaumeGomez replied:
You can only find `TyConst` when listing over an item generic arguments. An extern is simply a const variable defined in another crate.

GuillaumeGomez replied:
Take a look at `fn external_generic_args` in `clean/utils.rs`.

camelid replied:
Why do we need to track `Extern` separately from `Local`? It seems like you can get the value of the `Extern` just from the `DefId`, so can't we merge `Extern` and `Local`?

GuillaumeGomez replied:
The way they're displayed is both different (one use a `BodyId` whereas the other is forced to use its `DefId`. But ultimately, it'd be nice to be able to merge them. But I think it's out of the scope of this PR.

GuillaumeGomez commented:
@camelid @jyn514 Anything else to be done here? (Kinda worried about the coming rebases ^^')

camelid commented:
> (Kinda worried about the coming rebases ^^')

(What rebases?)

GuillaumeGomez commented:
Development in rustdoc is still going on so conflicts will very likely appear. Which is why I'd prefer to not keep this PR hanging around for too long.

jyn514 commented:
> Development in rustdoc is still going on so conflicts will very likely appear. Which is why I'd prefer to not keep this PR hanging around for too long.

I'm sorry, I can't commit to reviewing in any period of time. If you'd like it to be reviewed sooner you could maybe ping some other team members, but they may have other commitments too.

bors commented:
:umbrella: The latest upstream changes (presumably #82855) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

GuillaumeGomez commented:
r? @ollie27

jyn514 commented on src/librustdoc/json/conversions.rs:
```diff
@@ -87,59 +87,78 @@ impl JsonRenderer<'_> {
     }
 }
 
+crate trait FromWithTcx<T> {
+    fn from_tcx(f: T, tcx: TyCtxt<'_>) -> Self;
+}
+
+crate trait IntoWithTcx<T> {
+    fn into_tcx(self, tcx: TyCtxt<'_>) -> T;
+}
+
+impl<T, U> IntoWithTcx<U> for T
+where
+    U: FromWithTcx<T>,
+{
+    fn into_tcx(self, tcx: TyCtxt<'_>) -> U {
+        U::from_tcx(self, tcx)
+    }
+}
```
Why do these need to be traits? It seems simpler to just use free functions. Otherwise you'll have to redo this all again next time you want to add a parameter.

GuillaumeGomez replied:
I kept it as close as possible to the current code (which was using `From`). Otherwise I'll have to replace all the current trait implementations with free functions for all different types.

jyn514 replied:
Hmm, well I'm not convinced the traits are pulling their weight, but we can fix it in a follow-up to avoid unnecessary churn.

jyn514 reviewed (commented):
I think I'm ready to approve this with a couple nits fixed. I don't feel confident I've caught all the bugs, but hopefully the test suite will pick up anything that goes wrong.

GuillaumeGomez commented on src/librustdoc/clean/types.rs:
```diff
@@ -1978,9 +1979,55 @@ crate struct Static {
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+crate enum ConstantKind {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { expr: String },
```
:+1:

GuillaumeGomez commented:
Rebased to fix conflicts and updated doc comment as suggested by @jyn514. Now remains https://github.com/rust-lang/rust/pull/82873#discussion_r600151523.

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit 7b5908931606c02045482d0de9fbe2aa7f93eda1 has been approved by `jyn514`

<!-- @bors r=jyn514 7b5908931606c02045482d0de9fbe2aa7f93eda1 -->
<!-- homu: {"type":"Approved","sha":"7b5908931606c02045482d0de9fbe2aa7f93eda1","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 7b5908931606c02045482d0de9fbe2aa7f93eda1 with merge 3debe9acb8df363e0b5363f23f8218c2d7919904...
<!-- homu: {"type":"BuildStarted","head_sha":"7b5908931606c02045482d0de9fbe2aa7f93eda1","merge_sha":"3debe9acb8df363e0b5363f23f8218c2d7919904"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2198217221)
Approved by: jyn514
Pushing 3debe9acb8df363e0b5363f23f8218c2d7919904 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2198217221"},"merge_sha":"3debe9acb8df363e0b5363f23f8218c2d7919904"} -->
